### DISCUSSION TITLE
Vector Color Fill Loss

### DISCUSSION THREAD

**Discussion by loboere** [opentoonz/opentoonz#6580] [Q&A]
https://github.com/opentoonz/opentoonz/assets/15265895/31e69f50-bc77-4ecb-861f-6fb91be10d0b
opentoonz version 1.7.1
windows 8

**Comment by DarrenTAnims** [comment]
Can you give any more details. What does "Fill loss" mean? What did you do? What level type did you use? Without any detail, there's no help we can give.

**Comment by loboere** [comment]
@DarrenTAnims  It seems that you did not see the link that I left above. the video shows how the purple fill is lost when moving the vertices

**Comment by artisteacher** [comment]
Part of what's happening, is that anytime there is a change in node/handle position, OT is recalculating the fillable areas. Same thing happens when making adjustments to vectors with the selection tool, or adding strokes with the brush. I believe this is to help with cel shading.

I would guess that at first, you lost part of the fill because you moved the node in a way that made 2 fillable regions. When you pulled higher, it was recognized as a completely new region - and so ended up using the transparent style that was already there in the bg. Moving back, close to the original position, was considered another new region, so you didn't regain the color.

I'm not sure if this is intended, or not. There are definitely issues with how fillable regions are calculated. Vectors/svg need some attention.

**Comment by loboere** [comment]
inkscape does this well, I thought opentoonz would do the same

**Comment by RodneyBaker** [comment]
There are a number of 'rules' for color fill with vectors in Opentoonz that are different than with other vector based graphics programs.  Here's my general take although I wouldn't consider it official as I'm not likely correct or as precise as we could be:

- Vector fills are point based
-- This means that when a fill is assigned it is based on a calculation at the time of filling that assigns an index color from the color palette to a vector 'shape'.

- Vector 'shapes' have an assigned inside and outside related to the vector stroke/line.
-- When these lines/strokes are moved and the fill is recalculated confusion can arise where the process must determine what is inside versus what is outside

- Color Fills for vector shapes are applied only to eligible areas inside the same 'Group'.
-- This is a core rule of Color Fill with vectors and must strictly be adhered to

- Single (especially continuous) lines/strokes are particularly vulnerabke to losing fills when recalculation of fills cannot determine what is inside or outside of the line.  
-- In many cases rather than assign fills incorrectly it will drop the fill entirely.  The fill may be restored however when the lines are manipulated again and encounter the original point which the fill was based upon.  In some cases such as with manipulated shapes this can lead to color fills dropping out and returning over different frames of a sequence.

There are a few more 'rules' regarding vectors such as how gaps in lines effect fills but those are the ones that immediately come to mind that are the most significant.

Many programs that conserve the fills of vector shapes actually create a separate object that comprises the fill as opposed to the line work.  This is not the case with Opentoonz which almost completely relies upon the current linework to maintain the fill. 

Note that this discussion will likely be transferred to OT-Docs where it can be further considered.
Ongoing improvements to vector graphics is a topic of much interest to Opentoozn users and at present is proceeding slowly if only incrementally.

**Comment by RodneyBaker** [comment]
Will move this report to Opentoonz-Docs as it represents current state of vector fill in Opentoonz.
As we move forward we'll attempt to leverage vector technology to improve retention of color fill as control points are manipulated.
